### PR TITLE
feat: add custom routing support for ernie4_5_moe, and hunyuan_v1_moe

### PR DESCRIPTION
**Pull request by OnePunchMonk**
# Description
Added custom routing support for  ernie4_5_moe, deepseek_v2, hunyuan_v1_moe models as stated in 'src/axolotol/integrations/kernels/constants.py' at line 39 & issue #3151 

## How has this been tested?
Still untested

## AI Usage Disclaimer
Claude

## Summary by CodeRabbit

* **New Features**
  * Added end-user support for three MoE models: Ernie 4.5, DeepSeek V2, and HunYuan V1, each with tailored routing behavior.
  * Enabled flash-attention optimization for Ernie 4.5 where applicable.

* **Tests**
  * Added comprehensive tests validating routing outputs, normalization, group-limiting, bias/scaling effects, and error handling for the new MoE variants.

### CODE REVIEW

**Comment by OnePunchMonk** [comment]
@coderabbitai review

**Comment by OnePunchMonk** [comment]
@coderabbitai review

**Review by winglian — APPROVED** [review]
lgtm. good to go once @NanoCode012 approves

File: src/axolotl/loaders/patch_manager.py [hunk 1]
```diff
@@ -590,7 +590,8 @@ def _patch_llama_derived_model(self):
     def _apply_llama_flash_attn_patches(self, model):
         """Apply LLaMA-specific flash attention patches."""
         if (
-            self.model_config.model_type in ["llama", "llama4"]
+            self.model_config.model_type
+            in ["llama", "llama4", "ernie4_5", "ernie4_5_moe"]
```
Comment (NanoCode012) [inline src/axolotl/loaders/patch_manager.py:594]: Reasoning behind this?

File: src/axolotl/monkeypatch/lora_kernels.py [hunk 2]
```diff
@@ -172,6 +172,18 @@ def get_attention_cls_from_config(cfg: DictDefault) -> Type[nn.Module]:
...
+
+        return Ernie4_5Attention
+
+    if model_type == "ernie4_5_moe":
+        from transformers.models.ernie4_5_moe.modeling_ernie4_5_moe import (
+            Ernie4_5_MoeAttention,
+        )
+
+        return Ernie4_5_MoeAttention
+
```
Comment (NanoCode012) [inline src/axolotl/monkeypatch/lora_kernels.py]: Does it fail without? This is for edge case where the fallback doesn't work.

File: src/axolotl/utils/config/__init__.py [hunk 3]
```diff
@@ -208,7 +208,7 @@ def normalize_config(cfg):
     cfg.is_llama_derived_model = (
         (
             hasattr(model_config, "model_type")
-            and model_config.model_type in ["llama", "mllama_text_model"]
+            and model_config.model_type in ["llama", "mllama_text_model", "ernie4_5"]
```
Comment (NanoCode012) [inline src/axolotl/utils/config/__init__.py]: Is this correct?

File: src/axolotl/integrations/kernels/sonicmoe/routing.py [hunk 4]
```diff
@@ -276,3 +274,202 @@ def sigmoid_topk_routing(
...
+    # Compute logits in float32 and softmax
+    router_logits = F.linear(hidden_states.float(), gate.weight.float())  # [T, E]
+    router_probs = F.softmax(router_logits, dim=-1, dtype=torch.float32)  # [T, E]
+
+    if topk_method == "greedy":
+        topk_weights, topk_indices = torch.topk(router_probs, k=K, dim=-1, sorted=False)
+    elif topk_method == "group_limited_greedy":
+        # Guard: selected groups must contain enough experts for topk
+        group_size = num_experts // num_group
+        assert moe_block.topk_group * group_size >= K, (
```
Comment (NanoCode012) [inline src/axolotl/integrations/kernels/sonicmoe/routing.py]: Where is this check from? I don't see it upstream. We should also use ValueError instead of assert

File: src/axolotl/integrations/kernels/sonicmoe/routing.py [hunk 5]
```diff
@@ -57,17 +60,12 @@ def get_model_moe_config(model_type: str):
...
-    #     # (not n_group), gate is nn.Linear (not a router class).
-    #     return ..., ActivationType.SWIGLU, "gate"
-    # elif model_type in ("hunyuan_v1_moe",):
-    #     # Softmax→topk but gate structure differs: gate.wg (not gate.weight),
-    #     # top_k on block not gate, creates scatter routing matrix.
-    #     return ..., ActivationType.SWIGLU, "gate"
+    elif model_type in ("ernie4_5_moe",):
+        return softmax_bias_topk_routing, ActivationType.SWIGLU, "gate"
+    elif model_type in ("deepseek_v2",):
+        return softmax_group_limited_topk_routing, ActivationType.SWIGLU, "gate"
```
Comment (NanoCode012) [inline src/axolotl/integrations/kernels/sonicmoe/routing.py]: I left deepseekv2 as a note as it isn't really something folks are using. Maybe we don't need this until someone asks.

Comment (OnePunchMonk) [reply src/axolotl/integrations/kernels/sonicmoe/routing.py]: ok, have removed it

File: src/axolotl/integrations/kernels/sonicmoe/routing.py (hunk 4 above)
Comment (OnePunchMonk) [reply src/axolotl/integrations/kernels/sonicmoe/routing.py]: changed, this code is now irrelevant since deepseek v2 model wont be called

File: src/axolotl/monkeypatch/lora_kernels.py (hunk 2 above)
Comment (OnePunchMonk) [reply src/axolotl/monkeypatch/lora_kernels.py]: removed

File: src/axolotl/utils/config/__init__.py (hunk 3 above)
Comment (OnePunchMonk) [reply src/axolotl/utils/config/__init__.py]: yes, ernie 4.5 is a llama like model. It uses the same Rotary Positional Embeddings, RMSNorm, and SwiGLU activations.

File: src/axolotl/loaders/patch_manager.py (hunk 1 above)
Comment (OnePunchMonk) [reply src/axolotl/loaders/patch_manager.py:594]: Ref:
https://huggingface.co/docs/transformers/model_doc/ernie4_5
https://huggingface.co/docs/transformers/model_doc/ernie4_5_moe

Comment (OnePunchMonk) [reply src/axolotl/loaders/patch_manager.py:594]: The pages mention flash attention and Llama at its core, hence the inclusion
